pallets/phala/src/stakepool.rs

```diff
@@ -330,12 +371,59 @@ pub mod pallet {
 			);
 
 			pool_info.total_stake = pool_info.total_stake.saturating_add(amount.clone());
+			pool_info.free_stake = pool_info.free_stake.saturating_add(amount.clone());
+
+			// we have new free stake now, try handle the waitting withdraw queue
+			Self::try_handle_waitting_withdraw(&mut pool_info);
+
 			MiningPools::<T>::insert(&pid, &pool_info);
+
 			Self::deposit_event(Event::<T>::Deposit(pid, who, amount));
 
 			Ok(())
 		}
 
+		/// Deposits some funds from a pool
+		/// Note: there are two scenarios people may meet:
+		///     if the pool has free stake and and amount of the free stake greater or equal than withdraw amount
+		///     (e.g. pool.free_stake >= amount), the withdraw would take effect immediately.
+		///     else the withdraw would be queued and delay untill there have enough free stake in the pool.
+		/// Requires:
+		/// 1. The pool exists
+		#[pallet::weight(0)]
+		pub fn withdraw(origin: OriginFor<T>, pid: u64, amount: BalanceOf<T>) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let info_key = (pid.clone(), who.clone());
+			let mut user_info =
+				Self::staking_info(&info_key).ok_or(Error::<T>::StakeInfoNotFound)?;
+
+			ensure!(
+				&amount > &Zero::zero() && &user_info.amount >= &amount,
+				Error::<T>::InvalidWithdrawAmount
+			);
+
+			let mut pool_info = Self::mining_pools(&pid).ok_or(Error::<T>::PoolNotExist)?;
```

Comment:
Since `minin_pools().ok_or(Error)` is everywhere now, I suggest to write a helper function to simplify it:

```rust
fn ensure_pool(pid: u64) -> Result<PoolInfo, Error<T>> {
    Self::mining_pools(&pid).ok_or(Error::<T>::PoolNotExist)
}
```

Then you can write:

```rust
let mut pool_info = ensure_pool(pid)?;
```